File: src/engine/predictor.js
const bankProfiles = require('../config/bank_profiles.json');

class Predictor {
  /**
   * Predicts the probability that a pending/limbo transaction will resolve successfully on its own.
   * Pillar 1 - Step 3 in PDF specification.
   */
  predictResolutionProbability(txn) {
    const bankKey = txn.issuing_bank;
    const rail = txn.rail;
    const profile = bankProfiles[bankKey] || bankProfiles['HDFC'];
    const railConfig = profile.rails[rail] || profile.rails['UPI_AUTOPAY'];

    const baseSuccessRate = railConfig.success_rate;

    const debitTime = new Date(txn.debit_timestamp).getTime();
    const now = new Date().getTime();
    const ageSec = Math.max(0, (now - debitTime) / 1000);
    const expectedWindow = txn.expected_window_sec;

    // Time decay factor: if age exceeds expected window, confidence decays
    let timeFactor = 1.0;
    if (ageSec > expectedWindow) {
      const excessSec = ageSec - expectedWindow;
      // Decay lambda tuned per rail
      const lambda = rail === 'UPI_AUTOPAY' ? 0.005 : rail === 'CARD' ? 0.01 : 0.0001;
      timeFactor = Math.exp(-lambda * excessSec);
    }

    // Time-of-day adjustment (Off-peak bank batch processing 1 AM - 4 AM)
    const hour = new Date(txn.debit_timestamp).getHours();
    let todFactor = 1.0;
    if (hour >= 1 && hour <= 4) {
      todFactor = 0.90;
    }

    // Calculate final probability score
    let score = baseSuccessRate * timeFactor * todFactor;
    score = Math.min(0.99, Math.max(0.05, score));

    return parseFloat(score.toFixed(3));
  }

  /**
   * Scores the likelihood of a failed transaction succeeding if retried right now.
   * Pillar 2 - Step 1 in PDF specification.
   */
  scoreRetryWorthiness(txn) {
    const bankKey = txn.issuing_bank;
    const failureReason = txn.failure_reason || 'BANK_TIMEOUT';
    const retryCount = txn.retry_count || 0;

    const profile = bankProfiles[bankKey] || bankProfiles['HDFC'];
    const recoveryRates = profile.failure_recovery_rates || {};

    const baseRecovery = recoveryRates[failureReason] !== undefined
      ? recoveryRates[failureReason]
      : 0.50;

    // Multi-attempt decay: consecutive retries decrease likelihood of success
    const attemptDecay = Math.pow(0.65, retryCount);

    // Salary credit date boost (1st - 5th of month) for INSUFFICIENT_FUNDS
    const dayOfMonth = new Date().getDate();
    let salaryBoost = 1.0;
    if (failureReason === 'INSUFFICIENT_FUNDS' && (dayOfMonth <= 5 || dayOfMonth >= 28)) {
      salaryBoost = 1.35;
    }

    // Bank maintenance window check (1 AM - 3 AM)
    const currentHour = new Date().getHours();
    let maintenancePenalty = 1.0;
    if (currentHour >= 1 && currentHour <= 3) {
      maintenancePenalty = 0.40;
    }

    let score = baseRecovery * attemptDecay * salaryBoost * maintenancePenalty;
    score = Math.min(0.98, Math.max(0.02, score));

    return parseFloat(score.toFixed(3));
  }
}

module.exports = new Predictor();

File: src_py/predictor.py
"""
Project Limbo — Machine Learning & Statistical Prediction Engine
Calculates self-resolution probability scores P(success) for limbo payments
and evaluates retry-worthiness scores P(retry_success) for failed payments.
"""

import math
from datetime import datetime
from src_py.config import BANK_PROFILES

class Predictor:
    def predict_resolution_probability(self, txn: dict) -> float:
        """
        Pillar 1 - Step 3: Calculates probability P(self-resolution) for pending limbo payments.
        """
        bank_key = txn.get("issuing_bank", "HDFC")
        rail = txn.get("rail", "UPI_AUTOPAY")

        profile = BANK_PROFILES.get(bank_key, BANK_PROFILES["HDFC"])
        rail_config = profile["rails"].get(rail, profile["rails"]["UPI_AUTOPAY"])

        base_success_rate = rail_config["success_rate"]

        debit_time = datetime.fromisoformat(txn["debit_timestamp"].replace('Z', ''))
        now = datetime.utcnow()
        age_sec = max(0, (now - debit_time).total_seconds())
        expected_window = txn["expected_window_sec"]

        # Exponential time decay factor when age exceeds normal confirmation window
        time_factor = 1.0
        if age_sec > expected_window:
            excess_sec = age_sec - expected_window
            lambda_val = 0.005 if rail == "UPI_AUTOPAY" else (0.01 if rail == "CARD" else 0.0001)
            time_factor = math.exp(-lambda_val * excess_sec)

        # Off-peak bank batch processing adjustment (1 AM - 4 AM)
        hour = debit_time.hour
        tod_factor = 0.90 if (1 <= hour <= 4) else 1.0

        score = base_success_rate * time_factor * tod_factor
        score = min(0.99, max(0.05, score))

        return round(score, 3)

    def score_retry_worthiness(self, txn: dict) -> float:
        """
        Pillar 2 - Step 1: Scores likelihood of success if a failed transaction is retried right now.
        """
        bank_key = txn.get("issuing_bank", "HDFC")
        failure_reason = txn.get("failure_reason") or "BANK_TIMEOUT"
        retry_count = txn.get("retry_count", 0)

        profile = BANK_PROFILES.get(bank_key, BANK_PROFILES["HDFC"])
        recovery_rates = profile.get("failure_recovery_rates", {})

        base_recovery = recovery_rates.get(failure_reason, 0.50)

        # Attempt decay multiplier: consecutive retries decrease likelihood of success
        attempt_decay = math.pow(0.65, retry_count)

        # Salary credit date boost (1st - 5th of month) for INSUFFICIENT_FUNDS
        now = datetime.utcnow()
        day_of_month = now.day
        salary_boost = 1.35 if (failure_reason == "INSUFFICIENT_FUNDS" and (day_of_month <= 5 or day_of_month >= 28)) else 1.0

        # Bank maintenance window penalty (1 AM - 3 AM)
        maintenance_penalty = 0.40 if (1 <= now.hour <= 3) else 1.0

        score = base_recovery * attempt_decay * salary_boost * maintenance_penalty
        score = min(0.98, max(0.02, score))

        return round(score, 3)

predictor = Predictor()
